LiteDB/Client/Mapper/BsonMapper.GetEntityMapper.cs

```diff
@@ -0,0 +1,234 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+
+namespace LiteDB;
+
+public partial class BsonMapper
+{
+    /// <summary>
+    /// Mapping cache between Class/BsonDocument
+    /// </summary>
+    private readonly ConcurrentDictionary<Type, EntityMapper> _entities = new();
+
+    /// <summary>
+    /// Get property mapper between typed .NET class and BsonDocument - Cache results
+    /// </summary>
+    internal EntityMapper GetEntityMapper(Type type)
+    {
+        if (_entities.TryGetValue(type, out EntityMapper mapper))
+        {
+            return mapper;
+        }
+
+        using var cts = new CancellationTokenSource();
+        mapper = new EntityMapper(type, cts.Token);
+        if (_entities.TryAdd(type, mapper))
+        {
+            this.BuildEntityMapper(mapper);
+        }
+        cts.Cancel();
+        cts.Dispose();
+
+        return mapper;
```

Comment:
Oh yeah, another tryget and throw is missing in case add fails 😅

Do you think that would solve the issue?

Reply:
It should, though there is at least one other case where it might fail, but I'm really unfamiliar with the code so I might be wrong.
Let's say we have 2 classes `A` and `B`, each depending on each other:

```c#
class A
{
    public B Child { get; set; }
}

class B
{
    public A Parent { get; set; }
}
```

Now, let's say that we have 2 threads trying to resolve each class independently and concurrently (thread 1 resolves `A` while thread 2 resolves `B`). In this cases, they're both added to `_entities` successfully. Now here's a question. Does `BuildEntityMapper` at some point call `GetEntityMapper` for the property and attempts to wait for it? Because if so, then neither of both threads will be able to complete because thread 1 will wait for thread 2 while thread 2 will wait for thread 1.

Reply:
And something else I just thought. If `BuildEntityMapper` throws for any reason whatsoever, then it'll leave an incomplete `EntityMapper` there, making it completely unusable.

Reply:
Hey @vonzshik, thanks for your input. If you look at the changes, you will see, that on every de-/serialization, ``WaitForInitialization`` will be called. The nested/recursive serialization problem is addressed by creating an empty/uninitialized shell object first, which can be referenced by the other serializer. When the ``EntityMapper`` is requested for building the serializer, it shouldn't call ``WaitForInitialization``.

In addition, all unit tests passed, including the one that is about the recursive issue you described.

Reply:
I would appreciate your opinion on https://github.com/litedb-org/LiteDB/pull/2569 @vonzshik